file/file_instance.go

```diff
@@ -0,0 +1,325 @@
+	case segments.J1SegmentDescription:
+		f.J1Segments = append(f.J1Segments, s)
+	case segments.J2SegmentDescription:
+		f.J2Segments = append(f.J2Segments, s)
+	case segments.K1SegmentDescription, segments.K2SegmentDescription, segments.K3SegmentDescription,
+		segments.K4SegmentDescription, segments.L1SegmentDescription, segments.N1SegmentDescription:
+		f.Appendages[s.Description()] = s
+	}
+	return nil
+}
+
+// GetSegment returns single segment like as header, base, trailer k1, k2, k3, k4, l1, n1
+func (f *fileInstance) GetSegment(description string) segments.Segment {
+	switch description {
+	case segments.HeaderRecordDescription:
+		return f.Header
+	case segments.BaseSegmentDescription:
+		return f.Base
+	case segments.TrailerRecordDescription:
+		return f.Trailer
+	case segments.K1SegmentDescription, segments.K2SegmentDescription, segments.K3SegmentDescription,
+		segments.K4SegmentDescription, segments.L1SegmentDescription, segments.N1SegmentDescription:
+		if appendage, ok := f.Appendages[description]; ok {
+			return appendage
+		}
+	}
+	return nil
+}
+
+// GetSegment returns multiple segments like as j1, j2
+func (f *fileInstance) GetListSegments(description string) []segments.Segment {
+	switch description {
+	case segments.J1SegmentDescription:
+		return f.J1Segments
+	case segments.J2SegmentDescription:
+		return f.J2Segments
+	}
+	return nil
+}
+
+// GeneratorTrailer returns trailer segment that created automatically
+func (f *fileInstance) GeneratorTrailer() (segments.Segment, error) {
+	return nil, nil
+}
+
+// UnmarshalJSON parses a JSON blob
+func (f *fileInstance) UnmarshalJSON(data []byte) error {
+	f.reset()
+
+	var dummy map[string]interface{}
+	err := json.Unmarshal(data, &dummy)
+	if err != nil {
+		return nil
+	}
+
+	for key, record := range dummy {
+		buf, err := json.Marshal(record)
+		if err != nil {
+			return err
+		}
+
+		switch key {
+		case segments.BaseSegmentDescription:
+			err = json.Unmarshal(buf, f.Base)
+		case segments.HeaderRecordDescription:
+			err = json.Unmarshal(buf, f.Header)
+		case segments.TrailerRecordDescription:
+			err = json.Unmarshal(buf, f.Trailer)
+		case segments.J1SegmentDescription, segments.J2SegmentDescription:
+			var list []interface{}
+			err := json.Unmarshal(buf, &list)
+			if err != nil {
+				return nil
+			}
+			for _, subSegment := range list {
+				subBuf, err := json.Marshal(subSegment)
+				if err != nil {
+					return err
+				}
+				err = f.unmarshalApplicableSegments(key, subBuf)
+				if err != nil {
+					return err
+				}
+			}
+		default:
+			err = f.unmarshalApplicableSegments(key, buf)
+		}
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// MarshalJSON returns JSON blob
+func (f *fileInstance) MarshalJSON() ([]byte, error) {
+	buffer := bytes.NewBufferString("{")
+
+	value, err := json.Marshal(f.Header)
+	if err != nil {
+		return nil, err
+	}
+	buffer.WriteString(fmt.Sprintf("\"%s\":%s,", f.Header.Description(), string(value)))
```

Comment:
Will the struct tags not marshal properly here? We can ignore fields with `json:"-"` or override more MarhsalJSON funcs in structs. 

I don't like the `fmt.Sprintf(..)` calls here.

Reply:
updated the logic